src/input/type_base.cc

```diff
@@ -167,36 +109,7 @@ bool Array::ArrayData::finish()
 {
 	if (finished) return true;
 
-	if (p_type_of_values != 0)
-	{
-	    if (! was_constructed(p_type_of_values) ) return false;
-
-		if (dynamic_cast<const AbstractRecord *>(p_type_of_values) != 0)
-		{
-			AbstractRecord *ar = (AbstractRecord *)dynamic_cast<const AbstractRecord *>(p_type_of_values);
-			boost::shared_ptr<const TypeBase> type_copy = boost::make_shared<const AbstractRecord>(*ar);
-			type_of_values_ = type_copy;
-			p_type_of_values = 0;
-		}
-		else if (dynamic_cast<const Record *>(p_type_of_values) != 0)
-		{
-			Record *r = (Record *)dynamic_cast<const Record *>(p_type_of_values);
-			boost::shared_ptr<const TypeBase> type_copy = boost::make_shared<const Record>(*r);
-			type_of_values_ = type_copy;
-			p_type_of_values = 0;
-		}
-		else if (dynamic_cast<const Selection *>(p_type_of_values) != 0)
-		{
-			Selection *s = (Selection *)dynamic_cast<const Selection *>(p_type_of_values);
-			boost::shared_ptr<const TypeBase> type_copy = boost::make_shared<const Selection>(*s);
-			type_of_values_ = type_copy;
-			p_type_of_values = 0;
-		}
-		else if (dynamic_cast<const Array *>(p_type_of_values) != 0)
-		    xprintf(PrgErr, "Should not happen!\n");
-	}
-
-	return (finished = true);
+	return (finished = const_cast<TypeBase *>( type_of_values_.get() )->finish() );
```

Comment:
const_cast smells....
Possibly we can make finish method constant and make finished flag mutable.